### ISSUE TITLE
Failing conv+relu benchdnn tests with ACL backend in FP16 mode.

### ISSUE THREAD

**Issue by fadara01** [oneapi-src/oneDNN#1499]
# Summary
The recent integration of ACL F16 into oneDNN in [this commit](https://github.com/oneapi-src/oneDNN/commit/64d388c957a8edddccd59d1d14fa50dd32906560) caused the conv+relu tests to start failing
with ACL backend in F16 mode. The main reason for the failure is that the reference implementation does:
`F16(F32(src) * F32(alpa))` while ACL does `F16(alpha) * F16(src)`. The output of the convolution in ACL is in F16 and as a result the RELU post-op gets run using the [F16 kernels](https://github.com/ARM-software/ComputeLibrary/blob/main/src/cpu/kernels/activation/generic/sve/fp16.cpp).

I tried to fix this [here](https://github.com/oneapi-src/oneDNN/pull/1493) by relaxing the relative error, but the oneDNN API requires all post-ops to be executed in F32.
The issue can be (temporarily) resolved by disabling F16 ACL post-ops as in the patch attached.
[disable_f16_acl_postops.txt](https://github.com/oneapi-src/oneDNN/files/10014582/disable_f16_acl_postops.txt)
Note however, this means that conv+relu in F16 mode will start falling to reference, with the time to execute that layer (ignoring other operations such as reordering) going up from ~ 2.5 milli seconds to ~ 2.8 seconds (for the test case used to reproduce the issue).

# Version
oneDNN 2.7 and githash: 64d388c957a8edddccd59d1d14fa50dd32906560
# Environment

* CPU: aarch64
* OS version: Ubuntu 20.04
* Compiler version: GCC 10.3.0
* CMake version: 3.16.3

# Steps to reproduce
```fadara01@:~/oneDNN/build$ DNNL_VERBOSE=1 ./tests/benchdnn/benchdnn --conv --skip-impl=ref --allow-enum-tags-only=false --cfg=f16 --attr-post-ops=relu:0.1 mb1ic128ih52oc256oh52kh3ph1n"yolo.inf.fp16.ov.mb1*3"
onednn_verbose,info,oneDNN v2.7.0 (commit 64d388c957a8edddccd59d1d14fa50dd32906560)
onednn_verbose,info,cpu,runtime:OpenMP,nthr:16
onednn_verbose,info,cpu,isa:AArch64 SVE (256 bits)
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f32::blocked:abcd:f0 dst_f16::blocked:acdb:f0,,,1x128x52x52,8.94385
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f16::blocked:acdb:f0 dst_f32::blocked:abcd:f0,,,1x128x52x52,6.76294
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f32::blocked:abcd:f0 dst_f16::blocked:acdb:f0,,,256x128x3x3,7.44507
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f16::blocked:acdb:f0 dst_f32::blocked:abcd:f0,,,256x128x3x3,5.69385
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f32::blocked:a:f0 dst_f16::blocked:a:f0,,,256,0.059082
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f16::blocked:a:f0 dst_f32::blocked:a:f0,,,256,0.0488281
onednn_verbose,exec,cpu,convolution,gemm:acl,forward_training,src_f16::blocked:acdb:f0 wei_f16::blocked:acdb:f0 bia_f16::blocked:a:f0 dst_f16::blocked:acdb:f0,attr-post-ops:eltwise_relu:0.1 ,alg:convolution_direct,mb1_ic128oc256_ih52oh52kh3sh1dh0ph1_iw52ow52kw3sw1dw0pw1,2.50806
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f16::blocked:acdb:f0 dst_f32::blocked:abcd:f0,,,1x256x52x52,13.2168
[  13][DST][0:0:0:13] exp_f32:    -2.3 exp:-2.30078 got:-2.29883 diff:0.00195312 rdiff:0.000848896
[  21][DST][0:0:0:21] exp_f32:    -7.3 exp:-7.30078 got:-7.29688 diff:0.00390625 rdiff:0.000535045
[  31][DST][0:0:0:31] exp_f32:    -3.3 exp:-3.30078 got:-3.29883 diff:0.00195312 rdiff:0.000591716
[  41][DST][0:0:0:41] exp_f32:    -2.9 exp:-2.90039 got:-2.89844 diff:0.00195312 rdiff:0.000673401
[  46][DST][0:0:0:46] exp_f32:    -3.3 exp:-3.30078 got:-3.29883 diff:0.00195312 rdiff:0.000591716
[  52][DST][0:0:1:0] exp_f32:    -3.3 exp:-3.30078 got:-3.29883 diff:0.00195312 rdiff:0.000591716
[  59][DST][0:0:1:7] exp_f32:    -3.8 exp:-3.80078 got:-3.79883 diff:0.00195312 rdiff:0.000513875
[  67][DST][0:0:1:15] exp_f32:    -2.4 exp:-2.40039 got:-2.39844 diff:0.00195312 rdiff:0.00081367
[  77][DST][0:0:1:25] exp_f32:   -11.2 exp:-11.2031 got:-11.1953 diff:0.0078125 rdiff:0.00069735
0:FAILED (errors:61145 total:692224) __REPRO: --conv --skip-impl=ref --allow-enum-tags-only=false --cfg=f16 --attr-post-ops=relu:0.1 mb1ic128ih52oc256oh52kh3ph1nyolo.inf.fp16.ov.mb1*3
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:1 listed:0
total compute_ref: sum(s):0.81 
```

# Observed behavior
Test fails due to inaccuracies as `F16(src) * F16(alpha)` (what ACL does) `!= F32(F16(alpha) * F16(src))` (what reference does) 

# Expected behavior
Post-op is executed in F32 mode.

**Comment by nSircombe** [comment]
@lilh9598 - these BenchDNN test failures are a side effect of #1452. Although not something that the CI tests would have caught, they do show up when running a full test sweep, so a resolution is needed for v3.0 I believe.
Since setting the rel. error to avoid the problem (https://github.com/oneapi-src/oneDNN/pull/1493) isn't acceptable, the simplest action would be to disable ACL for the fp16 cases with post ops, as @fadara01 suggests (see attached patch - https://github.com/oneapi-src/oneDNN/files/10014582/disable_f16_acl_postops.txt). This does have a significant impact on the performance of these cases as they will fall back to reference.

A more desirable solution would be to ensure ACL does the postops into fp32 as oneDNN is expecting, but we don't have an implementation for this ready, and we would need to figure out if changes to ACL would be required.

As I say, disabling the cases with postops would be the most pragmatic thing to do until we have a better solution, but perhaps you'd be able to look at a less impactful solution?

**Comment by lilh9598** [comment]
@nSircombe Your analysis is reasonable, I don't have a better solution to avoid the problem. 

> Since setting the rel. error to avoid the problem (https://github.com/oneapi-src/oneDNN/pull/1493) isn't acceptable, the simplest action would be to disable ACL for the fp16 cases with post ops, as @fadara01 suggests (see attached patch - https://github.com/oneapi-src/oneDNN/files/10014582/disable_f16_acl_postops.txt). This does have a significant impact on the performance of these cases as they will fall back to reference.
> 
> A more desirable solution would be to ensure ACL does the postops into fp32 as oneDNN is expecting, but we don't have an implementation for this ready, and we would need to figure out if changes to ACL would be required.
> 

I think the best solution should be setting the rel. error, when users choose to use fp16 for inference, they know clearly that there will be a loss of accuracy.

**Comment by fadara01** [comment]
@lilh9598, unfortunately setting the error tolerance was not accepted by oneDNN, please see #1493. The oneDNN API specifies that post ops need to be executed in F32 mode in all the cases (even when everything else is in F16 mode). We need to adhere to that requirement, which rules out the "setting error tolerance" fix.

**Comment by jondea** [comment]
The ACL convolution does everything in FP16, including the accumulation, because this is what the hardware can accelerate effectively. I'm not sure if this violates the acc data type [defined here](https://github.com/oneapi-src/oneDNN/blob/master/tests/benchdnn/doc/driver_conv.md#precision-configurations)) or the [general numerical behaviour spec](https://oneapi-src.github.io/oneDNN/dev_guide_data_types.html#general-numerical-behavior-of-the-onednn-library) ([the reference convolution does all calculations in FP32 and only does a downcast before storing](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/ref_convolution.cpp#L216)). @dzarukin is this current ACL convolution behaviour acceptable? It does pass the correctness tests.

One thing to note is that the [benchdnn src is filled with moderately sized integers](https://github.com/oneapi-src/oneDNN/blob/master/tests/benchdnn/conv/conv.cpp#L132), which means the calculations in FP16 may not exhibit any round off errors. They fail with post ops with ReLU and alpha=0.1 because alpha isn't exactly representable in FP16.

One potential solution would be to allow the existing ACL approach if the [floating point mode is set to FP16](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html). By doing post ops in FP16 rather than FP32, we are in a sense implicitly casting to FP16. This would allow @lilh9598 to make use of the accelerated FP16 convolution while keeping the requirement for bit exact FP16 convolution+post ops. Does this work for the oneDNN team?

**Comment by dzarukin** [comment]
> The ACL convolution does everything in FP16, including the accumulation, because this is what the hardware can accelerate effectively.

I'm not sure this is going to fly with production frameworks when they start enabling fp16 data type on their side. In addition, this will mean that using different hardware will lead to pretty different results which affects user experience.

> One potential solution would be to allow the existing ACL approach if the [floating point mode is set to FP16](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html).

Fp16 fpmath mode still implies accumulation in fp32, just treat fp32 inputs as f16 instead. Using fp16 accumulation will definitely lead to significant precision loss.

I'll let @mgouicem and @vpirogov comment on expectations and usability side of this question.

**Comment by mgouicem** [comment]
I think this topic of numerical behavior for post-ops/scales was already discussed some time ago in #939 .
At the time the conclusion was to use ACL for the core computation and a separate kernel for post-ops.

Here are a few extra comments: 
- fp16 accumulation: this should be fine to use f16 accumulation for inference. We actually already have f16 accumulation for some kernels on Intel iGPU. However these kernels with f16 accumulation have to be disabled when oneDNN primitives will be used for f16 computations in frameworks, mostly because they can't be used in training. For example, PyTorch explicitely [requires f32 accumulation ](https://github.com/pytorch/pytorch/blob/ba605c3b0439fd5dfe062f42e60b990c88c061d4/aten/src/ATen/AccumulateType.h#L67).

- f16 scale / post-ops: so far the potential speedup from f16 post-ops was negligible compared to the risk of lower accuracy. To simplify validation and reduce risk of lower model accuracy, we decided to maintain all post-ops in f32. We could reassess this position but this would require some data on model accuracy before making a change in policy.

**Comment by fadara01** [comment]
> this should be fine to use f16 accumulation for inference
@mgouicem, Is it okay if we only do FP16 accumulation in forward inference?
@lilh9598 are you happy with the accuracy in this case?

**Comment by mgouicem** [comment]
@fadara01 I guess there are two venues to look at:
- guard the dispatching of those implementations with `prop_kind::forward_inference` as you mentioned
- guard the dispatching of those implementation with the `ONEDNN_ENABLE_WORKLOAD=INFERENCE` [build flag](https://oneapi-src.github.io/oneDNN/dev_guide_build_options.html).

Depending on which users are targetted by those f16 accumulation kernels, you might use one or the other options.
Though based on the following, I would recommend guarding the dispatching based on prop_kind.
- For Tensorflow/Pytorch, they always use `ONEDNN_ENABLE_WORKLOAD=TRAINING` and prop_kind::forward_training.
  There are a few exceptions where they use `prop_kind::forward_inference` when weights are constant or int8.
- For openVINO, they use `ONEDNN_ENABLE_WORKLOAD=INFERENCE` and `prop_kind::forward_inference` always
- For ONNX runtime, they use `ONEDNN_ENABLE_WORKLOAD=TRAINING`, and set prop_kind depending on a build option (always `forward_training` when training build, `forward_inference` otherwise).

**Comment by nSircombe** [comment]
@lilh9598 - do you have any opinions on using the `prop_kind` to filter cases and only dispatch to the ACL fp16 kernels for fwd inference, vs. making a build-time option (`ONEDNN_ENABLE_WORKLOAD=INFERENCE`) and not filtering by `prop_kind` at runtime.

...noting that: we've found many inference workloads often default to forward-training when they should/could use forward-inference; making it a build time option has implications for TF / PyTorch integration.

**Comment by mgouicem** [comment]
> making a build-time option (ONEDNN_ENABLE_WORKLOAD=INFERENCE) and not filtering by prop_kind at runtime.

Not that the build time option already exist. Though I don't think it is widely used as it is quite recent. Except openVINO, I don't think it is used anywhere.

> we've found many inference workloads often default to forward-training when they should/could use forward-inference

Could you share examples? This could be useful in case we need to raise that to the maintainers of these projects.

> making it a build time option has implications for TF / PyTorch integration.

TF and PyTorch integration use default option, so `ONEDNN_ENABLE_WORKLOAD=TRAINING`. Under that circumstance, it would mean that PT/TF would never dispatch the kernel with f16 accumulation. This is for the best I believe since FWK typically expect f32 accumulation (in particular PyTorch which explicitly sets all accumulation types to f32).

**Comment by lilh9598** [comment]
> @lilh9598 - do you have any opinions on using the prop_kind to filter cases and only dispatch to the ACL fp16 kernels for fwd inference, vs. making a build-time option (ONEDNN_ENABLE_WORKLOAD=INFERENCE) and not filtering by prop_kind at runtime.  

@nSircombe , Using the prop_kind to guard the dispatching of the ACL f16 kernels for fwd inference look good to me.

**Comment by vpirogov** [comment]
Considering that we are discussing changing the library semantics I'd suggest we define how this should work in RFC first. In the short term (i.e. for oneDNN v3.0 release) we need to either disable ACL fp16 implementations that are not conformant or make these opt-in using [experimental feature](https://oneapi-src.github.io/oneDNN/dev_guide_experimental.html) mechanism.

**Comment by vpirogov** [comment]
@mgouicem clarified that non-conformant part in this case is only fp16 post-op computations. Having fp16 accumulator in base operation is acceptable based on the current definition.

**Comment by igorsafo** [comment]
Closing the issue, since the fix was promoted (#1506). Thanks everyone!
